packages/flutter/lib/src/services/restoration.dart

```diff
@@ -0,0 +1,888 @@
+    @required RestorationId id,
+    @required RestorationBucket parent,
+    @required this.debugOwner,
+  }) : assert(id != null),
+       assert(parent != null),
+       assert(parent._rawChildren[id.value] != null),
+       _manager = parent._manager,
+       _parent = parent,
+       _rawData = parent._rawChildren[id.value] as Map<dynamic, dynamic>,
+       _id = id;
+
+  static const String _childrenMapKey = 'c';
+  static const String _valuesMapKey = 'v';
+
+  final Map<dynamic, dynamic> _rawData;
+
+  /// The owner of the bucket that was provided when the bucket was claimed via
+  /// [claimChild].
+  ///
+  /// The value is used in error messages.
+  final Object debugOwner;
+
+  RestorationManager _manager;
+  RestorationBucket _parent;
+
+  /// The [RestorationId] under which the bucket is currently stored in the
+  /// parent of this bucket (or wants to be stored if it is currently
+  /// parent-less).
+  ///
+  /// This value is never null.
+  RestorationId get id => _id;
+  RestorationId _id;
+
+  Map<dynamic, dynamic> get _rawChildren => _rawData.putIfAbsent(_childrenMapKey, () => <dynamic, dynamic>{}) as Map<dynamic, dynamic>;
+  Map<dynamic, dynamic> get _rawValues => _rawData.putIfAbsent(_valuesMapKey, () => <dynamic, dynamic>{}) as Map<dynamic, dynamic>;
+
+  /// Called to signal that this bucket and all its descendants are no longer
+  /// part of the current restoration data and must not be used anymore.
+  ///
+  /// Calling this method will drop this bucket from its parent and notify all
+  /// its listeners as well as all listeners of its descendants. Once a bucket
+  /// has notified its listeners, it must not be used anymore. The bucket must
+  /// be [dispose]d and replaced with a new bucket.
+  ///
+  /// As an example, the [RestorationManager] calls this method on its root
+  /// bucket when it has been asked to restore a running application to a
+  /// different state. At that point, the data stored in the current bucket
+  /// hierarchy is invalid and will be replaced with a new hierarchy generated
+  /// from the restoration data describing the new state. To replace the current
+  /// bucket hierarchy, [decommission] is called on the root bucket to signal to
+  /// all owners of buckets in the hierarchy that their bucket has become
+  /// invalid. In response to the notification, bucket owners must [dispose]
+  /// their buckets and claim a new bucket from the newly created hierarchy. For
+  /// example, the owner of a bucket that was originally claimed from the
+  /// [RestorationManager.rootBucket] must dispose that bucket and claim a new
+  /// bucket from the new [RestorationManager.rootBucket]. Once the new bucket
+  /// is claimed, owners should restore their state according to the data stored
+  /// in the new bucket.
+  ///
+  /// A call to [decommission] must always be followed by a call to [dispose].
+  void decommission() {
+    _parent?._dropChild(this);
+    _parent = null;
```

Comment:
nit: This would be more efficient as:

```dart
if (_parent != null) {
  parent._dropChild(this);
  parent_ = null;
}
```

right?

Reply:
Why have this method _and_ dispose?

Reply:
(IOW why not just fold it into dispose)

Reply:
Dispose and decommission are different. Decommission is a notification that this entire bucket subtree is getting replaced with a new bucket hierarchy. All the children are informed about this so the owner of them can dispose them and claim a new bucket.

Dispose simply removes the information of this bucket from the hierarchy. The child bucket continue to have a live of their own and the owners of those child buckets may get them adopted by other parents.

Decommission = notification down the tree to stop using any bucket in this hierarchy.
dispose = delete the data from this bucket, child buckets can continue to live if they want.